kernel/src/debug.rs

```diff
@@ -285,22 +321,49 @@ macro_rules! debug_gpio {
 ///////////////////////////////////////////////////////////////////
 // debug! and debug_verbose! support
 
+///  A trait that should be implemented for debug writers other than the provided [`UartDebugWriter`].
+///
+///  This can be used for example to implement synchronous or
+/// unbuffered writers, such as in-memory logs, memory mapped "endless" FIFOs etc.
+pub trait DebugWriter {
+    /// Write bytes to output
+    /// The `overflow` slice is used as a message to be appended to the end of the available
+    /// buffer if it becomes full.
+    fn write(&self, bytes: &[u8], overflow: &[u8]) -> usize;
+
+    /// Available length of the internal buffer
+    fn available_len(&self) -> usize {
+        // Default implementation is to return maximum (for synchronous/unbuffered implementations)
+        usize::MAX
+    }
+    /// Publish bytes from the internal buffer to the output
+    fn publish_bytes(&self) -> usize {
+        // Default implementation is to do nothing
+        0
+    }
```

Comment:
What does the return value indicate? It is legal for an implementation to "publish" bytes before this method is called? Should this really be called "flush"?